### PR TITLE
feat: setup hatchling build-system and use @anywidget/react

### PR DESCRIPTION
**Pull request by manzt**
Some additional touch-ups for the repo. This is probably how I would package things.

- Uses `hatchling` build-system with plugin to run `npm build` when packaging the repo (`python -m build`).
- Removes build artifact (`static/widget.js`) and .gitignores this file since it is produced by the build system.
- Uses the `@anywidget/react` library I published yesterday to remove additional boilerplate.

### CODE REVIEW

**Comment by kolibril13** [comment]
I've now done `brew install hatch` and then did the following:

```
# clone repo
cd js
npm run install
npm run dev

cd ..

# then as I read in https://hatch.pypa.io/latest/environment/

hatch env create 
hatch shell

# then I get:

source "/Users/jan-hendrik/Library/Application Support/hatch/env/virtual/qr-code-widget/luncB7zc/qr-code-widget/bin/activate"
jan-hendrik@JanHendriksAir anywidget-react-vite-test % source "/Users/jan-hendrik/Library/A
pplication Support/hatch/env/virtual/qr-code-widget/luncB7zc/qr-code-widget/bin/activate"

(qr-code-widget) jan-hendrik@JanHendriksAir anywidget-react-vite-test % 


 now I go to VS Code and type in the command palette (Shift + Command + P):
> Python: Select Interpreter
> /Users/jan-hendrik/Library/Application Support/hatch/env/virtual/qr-code-widget/luncB7zc/qr-code-widget/bin/python

then I open the notebook in VS Code, go to the kernel picker and select the `qr-code-widget`

now I run the notebook, and get:
```
image

Nice!
Now I want to build the package, so what I do is
```
(qr-code-widget) jan-hendrik@JanHendriksAir anywidget-react-vite-test % hatch build
# question here: Should I run that command while I'm in the (qr-code-widget) environment, or does that not matter?

now  dist/qr_code_widget-0.0.0-py2.py3-none-any.whl was created.

# lets test this in another evironment in a completly empty new project:
python3.11 -m venv .venv && source .venv/bin/activate
pip install qr_code_widget-0.0.0-py2.py3-none-any.whl
pip install jupyter lab
jupyter lab
```
and now I get:
image
now I get this error.
What did I do wrong?

**Comment by kolibril13** [comment]
just tried again with 
```
npm run build

# now the widget.js is about 6 times smaller in in size then before

hatch build

# and again empty new project:
python3.11 -m venv .venv && source .venv/bin/activate
pip install qr_code_widget-0.0.0-py2.py3-none-any.whl
pip install jupyter lab
jupyter lab
```
but still same error: SyntaxError: Invalid regular expression flags

**Comment by manzt** [comment]
oops, ya. I had a typo in the pyproject.toml. It should be "artifacts" not "artifact". The `widget.js` wasn't ending up in the final build.

**Comment by manzt** [comment]
BTW, you don't need `hatch` CLI to work on this project. All the project metadata is PEP compliant:

For development:

- Clone the repo
- Create a virtual env and install in development mode

```sh
python3 -m venv .venv 
source .venv/bin/actviate
pip install -e ".[dev]" # This will run npm install & npm run build for you the first time
```

- Start esbuild dev server (`cd js && npm run dev`) and open `hello.ipynb`

To publish, make sure you have `build` installed (`pip install build`)

```sh
python -m build .
```

and upload `dist` with twine

**Comment by manzt** [comment]
Now, I prefer `hatch` CLI for all my projects now because it is basically an alias for venv/build/twine. Same flow as above with `hatch`:

For development:

- Clone the repo
- Create a vritual environment and install in development mode

```sh
hatch shell # creates virtual env and activates it
```

- Start esbuild dev server (cd js && npm run dev) and open hello.ipynb

To publish:

- `hatch build`
- `hatch publish`

**Comment by kolibril13** [comment]
> oops, ya. I had a typo in the pyproject.toml. I 

awesome, thanks for this super quick fix.
I can confirm that it works now! 🎉

**Comment by kolibril13** [comment]
So for **development** I now have
```
python3.11 -m venv .venv && source .venv/bin/activate
pip install -e ".[dev]" 

# oh wow, that already created scr/qr_code_widgt/widget.js

# and then

cd js && npm run dev

# then open VS Code notebook, where the kernel is already automatically set, that's awesome!
```
I think I prefer this over the hatch environment approach because this way I always have the .venv always in one place, and it has also always the same name and same relative path. No pain to find the right environment in the vs code notebook kernel picker. And also when I delete a project, I can also be sure the environment got deleted as well. 

Regarding **publishing**:

> BTW, you don't need hatch CLI to work on this project

Does that mean that the workflow is also possible without hatch being installed?

When hatch would have to be installed eighter way, why then installing twine to upload the dist folder and not simply use `hatch publish`

regarding **this pr**:

is there something you want to add, or can it be merged already?

File: .gitignore [hunk 1]
```diff
@@ -24,4 +24,6 @@ dist-ssr
 *.sw?
 
 __pycache__
-.ipynb_checkpoints
\ No newline at end of file
+.ipynb_checkpoints
+
+src/qr_code_widget/widget.js
```
Comment (kolibril13) [inline .gitignore:29]: should we generalize this a bit more?

e.g. to 
```
src/*.js
src/*.css
```

**Comment by kolibril13** [comment]
also, I can confirm that multi-widget export works now as well!
image

**Comment by manzt** [comment]
> I think I prefer this over the hatch environment approach because this way I always have the .venv always in one place, and it has also always the same name and same relative path. No pain to find the right environment in the vs code notebook kernel picker. And also when I delete a project, I can also be sure the environment got deleted as well.

Sure thing. FWIW, you can configure where hatch creates its virtual environments in its [global config](https://hatch.pypa.io/latest/blog/2022/10/08/hatch-v160/#virtual-environment-location). For example, I set:

```toml
[dirs.env]
virtual = ".venv"
```

in my config: `~/Library/Application\ Support/hatch/config.toml`.

> Does that mean that the workflow is also possible without hatch being installed?

Yes, exactly! `python -m build && twine upload dist/`

File: .gitignore (hunk 1 above)
Comment (manzt) [reply .gitignore:29]: Yah, that would be great.

**Comment by kolibril13** [comment]
> Sure thing. FWIW, you can configure where hatch creates its virtual environments in its [global config](https://hatch.pypa.io/latest/blog/2022/10/08/hatch-v160/#virtual-environment-location). For example, I set:
> 
> ```toml
> [dirs.env]
> virtual = ".venv"
> ```
> 
> in my config: `~/Library/Application\ Support/hatch/config.toml`.

Nice! In case that I could really work with hatch in my .venv folder like I'm used to with python venv, I would switch :)
Just tried what you mention here, but now there's a subfolder .venv/qr-code-widget/bin and not .venv/bin , so the VS Code notebook kernel picker does not auto-find the venv there, that's unfortunate. Is there a way to remove that extra sub-dir?

**Comment by manzt** [comment]
> also, I can confirm that multi-widget export works now as well!

That's great! The `@anywidget/react` stuff is kind of experimental, but I think I like the pattern.

**Comment by manzt** [comment]
> Is there a way to remove that extra sub-dir?

hmm, I'm sorry I don't know! I don't develop using VS Code.

**Comment by manzt** [comment]
> is there something you want to add, or can it be merged already?

nope. merge away!

**Comment by kolibril13** [comment]
>  That's great! The @anywidget/react stuff is kind of experimental, but I think I like the pattern

Amazing, I'll start to incorporate this into https://github.com/kolibril13/ipymafs soon :)

> Maybe ask a question to hatch?

good idea, I'll do that!


> nope. merge away!

I'll do that, thanks for this journey! 👍🏼

**Comment by kolibril13** [comment]
I think we should rename the repo now, as  anywidget-react-vite-test is not up-to-date anymore.
How about anywidget-minimal-example-for-react?

**Comment by manzt** [comment]
or `anywidget-react-template` / `anywidget-minimal-react-template`?

**Comment by manzt** [comment]
Other thing would be to add a minimal readme with development instructions.

**Comment by kolibril13** [comment]
Sounds good! 
I prefer 
anywidget-minimal-react-template
as this would be easier to have it in a list with potential other minimal examples like
anywidget-minimal-svelte-template

just started to migrate the template to my project:
image
Maybe a cookiecutter would be a good idea here, as there are quite a few things that have to be changed.

>  Other thing would be to add a minimal readme with development instructions.

I can make a pr for that!

**Comment by manzt** [comment]
> Maybe a cookiecutter would be a good idea here, as there are quite a few things that have to be changed.

Right. I kind of dislike cookiecutters (whole reason we got here in the first place!), and worry about keeping up to date. Mostly worry about maintenance of templates over time and things left out in the wild that quickly go out of date. Build tools will always change, but very unlikely that ESM will change over time.

With that said, one thing I'd been thinking about is some templates via an _npm_-based CLI like vite: https://vitejs.dev/guide/#scaffolding-your-first-vite-project. 

```sh
npx create anywidget@latest --template react
```

the nice thing about this is that in order to scaffold the project, users need to have npm installed.

**Comment by kolibril13** [comment]
I totally see your cookie cutter concern. 
A minimal example is deftly easier to maintain and with an instruction how to work with it, people will be able to figure out stuff on their own. 

Also, one more point regarding folder nesting:
Currently, we have /src/ which is just a sub-folder that seems quite unnecessary to me.
/src/qr_code_widget
can we remove that and have only
/qr_code_widget
instead?

**Comment by manzt** [comment]
> A minimal example is deftly easier to maintain and with an instruction how to work with it, people will be able to figure out stuff on their own.

Sure makes sense! Thanks for all your thoughts on improving the react with anywidget.

> Also, one more point regarding folder nesting:

This is just a convention (`src-layout`) that I've adopted in most my python projects. It seems to be adopted in recommended python packaging now. Some discussion about benefits [here](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/).

**Comment by kolibril13** [comment]
> Sure makes sense! Thanks for all your thoughts on improving the react with anywidget.

It's a pleasure, I can only say that your vision of bringing a smooth widget development experience to everyone is really inspiring, and I can already foresee a lot of useful projects on the horizon 🌅

> This is just a convention (src-layout) that I've adopted in most my python projects

Ah, nice!
With the src-layout people might get into this error (happened to me right now :D )
image
personally I'd prefer the flat layout, but as long as the src layout is  documented somewhere, I'm happy with that as well!

**Comment by kolibril13** [comment]
> With the src-layout people might get into this error (happened to me right now :D )

next pitfall:
When running `pip install -e ".[dev]"` _before_ choosing a new name which is not the name of the minimal example, you have to run `pip install -e ".[dev]"`  after the renaming again.

**Comment by kolibril13** [comment]
oh, and one more question:

Should 
"widget.js"
"widget.css"
be in the same folder like
`__init__.py` ?

or should they be in a "static" subfolder like
"static/widget.js"
"static/widget.css"

**Pull request merged**